Add basic support for CSS grid prefixing

etripier opened the pull request:
This PR implements very basic style prefixing for CSS grid. This is the first step to solving https://github.com/rofrischmann/inline-style-prefixer/issues/173.

First, here's what I added:

- Prefixing for `display: grid | inline-grid`
- Prefixing / expansion for `grid-column | grid-column-start | grid-column-end | grid-row | grid-row-start | grid-row-end`
- Prefixing for `align-self` and `justify-self`
- Prefixing for `grid-template-columns | grid-template-rows`

Here's what I did _not_ add (beyond the properties not listed above, obviously):

- Support for `span` values
- `repeat` conversion (e.g., `repeat(2, 1fr)` -> `(1fr)[2]`)
- Auto-placement support
- Support for negative values in `grid-column` and `grid-row` (see caveats at https://css-tricks.com/css-grid-in-ie-css-grid-and-the-new-autoprefixer/)

This provides basic support for IE10/11 and early versions of Edge. The rest is difficult to add given the limited ability of this program to set styles on child elements (a general limitation of style composability imposed by React). I think it's possible to convert the `repeat` syntax and add basic `span` support, but that looks like it may be reaching beyond the scope of this program.

Ah! One last note: I noticed that the grid values in `propertyMap.js` may be generating invalid styles. Should I try to trim that down?

etripier commented:
@garrettberg @ljharb @lencioni @calinoracation Hey there! Would you take a look whenever you have a chance? 🙏

lencioni commented on modules/plugins/grid.js:
```diff
@@ -0,0 +1,95 @@
+/* @flow */
+
+function isSimplePositionValue(value: any) {
+  return typeof value === 'number' && !isNaN(value)
```
Can we use `Number.isNaN` instead of this function? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN

ljharb replied:
why? `isNaN` is available in more places, and the `typeof` ensures it'll behave identically.

lencioni replied:
Yeah we came full circle on this one

etripier commented on modules/__tests__/createPrefixer-test.js:
```diff
@@ -431,5 +431,119 @@ describe('Static Prefixer', () => {
+          const output = {
+            gridTemplateColumns: '1fr auto',
+            msGridColumns: '1fr auto',
+            msGridTemplateColumns: '1fr auto', // Not a valid property
```
The list of properties defined [here](https://github.com/rofrischmann/inline-style-prefixer/blob/master/modules/generator/maps/propertyMap.js#L126) causes this prefix to be automatically generated, even if it isn't valid (to the best of my knowledge). I don't have full context on why this is generated, but I think it could be removed - just waiting to hear back for confirmation.

etripier commented:
@lencioni Hey Joe, thanks for the comments! I went ahead and addressed them.

lencioni commented on modules/plugins/grid.js:
```diff
@@ -1,10 +1,10 @@
 
 function isSimplePositionValue(value: any) {
-  return typeof value === 'number' && !isNaN(value)
+  return typeof value === 'number' && !Number.isNaN(value)
```
Actually, I wonder if `Number.isFinite` is what you want here.

Also, I noticed that IE does not support either, so maybe what you originally had would be better in the end?

etripier replied:
@lencioni I went back to `isNaN` and removed the use of `Set` because I don't think IE11 supports constructing one [from an iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#Browser_compatibility).

ljharb commented on modules/plugins/grid.js:
```diff
@@ -0,0 +1,95 @@
+  return typeof value === 'number' && !isNaN(value)
+}
+
+const alignmentValues = ['center', 'end', 'start', 'stretch']
```
It definitely shouldn't matter, but an object would be fine too i suppose.

etripier commented:
@rofrischmann Hey Robin! I'm trying to add basic support for CSS grid. Would you please take a look when you have a chance?

rofrischmann commented:
Sure! Ill check it once I wake up tomorrow! Is it ready yet? Wanna get it merged and releasd asap :p

etripier commented:
It's ready! It doesn't fully cover the grid spec, but it adds enough basic compatibility that you can make things work on IE10/11. I might try to add more advanced features in other PRs.

etripier commented:
@rofrischmann Hey there, would you please take a look when you have time? I think this is ready to go.

rofrischmann commented:
About the `propertyMap.js`, did you remove that? After all,  it should just add prefixes to some values, so it won't break anything. But, it might be unnecessary as well.

etripier commented:
@rofrischmann I think we can probably take it out. I took a look at the [older IE spec](https://www.w3.org/TR/2011/WD-css3-grid-layout-20110407/#defining-grid-rows-columns-and-lines) and some MDN data and came up with this:

```
gridTemplateColumns - `-ms-grid-columns` is used instead
gridTemplateRows - `-ms-grid-rows` is used instead
gridTemplateAreas - not supported in IE
gridTemplate - not supported in IE
gridAutoColumns - `-ms-grid-columns` is used instead
gridAutoRows - `-ms-grid-rows` is used instead
gridAutoFlow - not supported in IE
grid - not supported in IE
gridRowStart - `-ms-grid-row` is used instead (not a 1-1 match)
gridColumnStart - `-ms-grid-column` is used instead (not a 1-1 match)
gridRowEnd - `-ms-grid-row-span` is used instead (not a 1-1 match)
gridRow - not a 1-1 match, implemented in the PR
gridColumn - not a 1-1 match, implemented in the PR
gridColumnEnd - `-ms-grid-column-span` is used instead (not a 1-1 match)
gridColumnGap - not supported in IE
gridRowGap - not supported in IE
gridArea - not supported in IE
gridGap - not supported in IE
```

lencioni commented:
:shipit:

rofrischmann commented:
Sorry for the delay. I had some unpredictable incidents in my family that required my full attention.
I'll try to merge and release this one asap. Probably on monday or tuesday.

rofrischmann commented:
Weeeh, its out at 5.1.0 :P

etripier commented:
Thank you so much! 🙏

kevinweber commented on modules/__tests__/createPrefixer-test.js:
```diff
@@ -431,5 +431,120 @@ describe('Static Prefixer', () => {
+        })
+
+        it('should not expand non-numerical values', () => {
+          const input = { gridRow: '2 / span 3' }
```
@lencioni What's the reason why we wouldn't want `2 / span 3` to expand?
